### ISSUE TITLE
Extract json from multipart request in endpoint

### ISSUE THREAD

**Issue by PhyberApex** [home-assistant/appdaemon#821]
A lot of webhook integrations (Emby / Plex) user multipart format for their webhooks. Currently Appdaemon fails with a json decode error. It would be great if it could be fixed so the request is looked at and if it is multipart it checks if there is a json part to it. As I do not have enough experience with python development I think I would fail horribly at trying to implement this.

Any chance this gets looked at?

~Cheers

**Comment by Odianosen25** [comment]
Hello @PhyberApex,

Can you kindly give more details what you trying to do, and how the webhook comes into play?

Also what version of AD you using, and where exactly in AD the error is seen.

Regards

**Comment by PhyberApex** [comment]
Sure as for context: I am using the appdaemon in docker which runs on version 3.0.5 currently. I am using the "register_endpoint" API function to register an endpoint called "emby_webhook_receiver". This endpoint gets called from my Emby server via a Webhook. This webhook is encoded with as multipart which contains one part in json. If needed I can attach what the request exactly looks like if that would help, but it's just a regular multipart request. Every time this webhook gets called I get an error in the AD logs that says: "INFO AppDaemon: API Call to emby_webhook_receiver: status: 400 JSON Decode Error". Which I believe stems from those lines https://github.com/home-assistant/appdaemon/blob/a18d26b26256c2edddb3813fccb6d0262c8aea0e/appdaemon/http.py#L738-L741

It seems like this is not able to handle multipart requests. Does this help?

~Cheers

**Comment by PhyberApex** [comment]
Here is the test webhook request from emby.
```
POST / HTTP/1.1
User-Agent: Emby Server/4.3.0.30
Accept-Encoding: deflate
Content-Type: multipart/form-data; boundary="e9bb8e51-f31d-4962-9c05-019577efba29"
Content-Length: 267
Host: localhost:8099

--e9bb8e51-f31d-4962-9c05-019577efba29
Content-Type: application/json; charset=utf-8
Content-Disposition: form-data

{"Event":"system.webhooktest","Server":{"Name":"Emby Server","Id":"365a5ea81ff7492f8d8d3197fdf47ec0"}}
--e9bb8e51-f31d-4962-9c05-019577efba29--
```

~Cheers

**Comment by Odianosen25** [comment]
Apologies I am just responding @PhyberApex ,

What is the response like that is to be sent back?

Besides the api end point isn't setup for something like that. I would expect you use a plugin kind of setup for what you want.

Also there is a PR in place to allow for apps to register web routes, look at it here #815. Will it solve your issue when its available? That can support multipart. And if it doesn't, do let me know what extra stuffs could be added to allow for your use case.

Regards

**Comment by PhyberApex** [comment]
The response that is send back doesn't actually matter as long as it has the http status of 200 (OK). I don't think the PR you mentioned would change any of that. Maybe I should make an example to better get my point across. Imagine the following setup:
```
Emby Client <-> Emby Media Server
                         ^
                         | Webhook (multipart with json part which includes the player id of the emby client)
                         v
                         AppDaemon <-> HomeAssistant <-> Livingroom lights
```
If I now press play on my Emby Client I want the lights in the living room to dim.

The emby Media server has a plugin for exactly such use cases called the webhook plugin. It makes a HTTP Post request (unfortunately multipart) to any given URL for certain events (like playing media).

So my solution was as follows: I use appdaemon to register an endpoint called "emby_webhook_receiver" and point the webhook plugin to that url (host:port/api/appdaemon/emby_webhook_receiver). I then decide what action to take in the callback of that registered endpoint (dim the lights, if the "player id" in the sent json payload matches. As of now this is not possible due to the post request being multipart (It worked before, but the webhook switched from pure json to multipart to include images in the future).

So for my use case to work the handling of requests for registered endpoints had to be made multipart supported. If I am not mistaken this is possible with aiohttp like mentioned [here](https://docs.aiohttp.org/en/stable/multipart.html). This has been a problem within home assistant itself before as can be seen [here](https://community.home-assistant.io/t/plex-webhooks-and-home-assistant/12090/38).

Hope this helps to clarify what kind of feature I am requesting.

~Cheers

**Comment by Odianosen25** [comment]
Hello @PhyberApex,

Thanks for the extra explanation. Well it seems it shouldn't be an issue adding it to AD, but its presently not in the pipeline. I will flag this as enhancement, so we can look into it later. It could be done with a plugin, but since its something that could potentially easily be added into the http component, I will suggest to the other devs we do it that way, but I can't guarantee that.

So based on the above, unfortunately I don't think its something we will be doing anytime soon; at least until AD 4.0 has been officially released. If you have anyway you manage it now, you can share it and we can look into merging it, that will be a faster route to getting it in.

Kind regards

**Comment by PhyberApex** [comment]
Okay so just to make sure I understood correctly. You would be open to a PR about this? I might take a shot at it then this weekend or try the plugin way. As I am not proficient in python it might take me some trial and errors before I get anything that actually works. I will report back either way tho. If I wanted to do a PR is there anything I have to take into account like unit tests or something?

~Cheers

**Comment by Odianosen25** [comment]
Yes we are open to a PR, and if having any difficulty, we are also willing to help.

We have not implemented any units tests yet, but there is a PR for that also which haven't been merged. So if you have anything, just submit and we will test it ourselves and look into if it can be merged or not.

Regards

**Comment by racoonx2p** [comment]
I would be glad for this feature as well. I'm trying to setup Interactive Components in Slack with Appdaemon **register_endpoint** function but response is multipart request.  I can see same error in log:

`API Call to my_api_link: status: 400 JSON Decode Error`

I can do it with home assistant Automation and Scripts as advised in this link https://community.home-assistant.io/t/how-to-get-actionable-notifications-using-slack/145035. But I would prefer appdaemon for this.

 @PhyberApex please let me know if you find workaround. Thanks.
